111-use-buildah-instead-of-kaniko.md

```diff
@@ -0,0 +1,239 @@
+An automated warning about deprecation will be added inside the `.status` section of the `KafkaConnect` CR in case it is used as for all other deprecated fields.
+Instead of `additionalKanikoOptions`, users should use `additionalBuildOptions`, but not `additionalPushOptions`.
+We will not split the Kaniko's additional options to two groups, and we will keep everything configured in one.
+That's because Kaniko doesn't have the build and push phases split, but everything is done in one command.
+The difference between the build and push options in Kaniko is that the `-pull` suffix is applied for options used only for the build phase.
+
+Buildah, in the other hand, has these phases split and has its own set of options for both.
+In order to allow specifying different options and their values for both of these phases, we will add two fields.
+Additionally, if user will use Kaniko and specify both `additionalKanikoOptions` and `additionalBuildOptions`, `additionalBuildOptions` takes precedence.
+
+For Buildah, we will check only the `additionalBuildOptions` and `additionalPushOptions` fields.
+In case that user will specify `additionalKanikoOptions` with Buildah enabled feature gate, two conditions with warnings will be added into `.status` section of `KafkaConnect` CR - one about deprecation and one about `additionalKanikoOptions` being ignored, as Buildah is used.
+
+As in case of Kaniko, Buildah will have also some allowed options that can be specified for the `build` and `push` phases:
+
+- `additionalBuildOptions`
+  - `--authfile` - path to the authentication file (usually created using the buildah login).
+  - `--cert-dir` - path to directory with certificates that should be used.
+  - `--creds` - the `[username[:password]]` to use to authenticate with the registry if required.
+  - `--decryption-key` - the `[key[:passphrase]]` to be used for decryption of images.
+  - `--retry` - number of times to retry in case of failure during image pull (the base one).
+  - `--retry-delay` - duration between retry attempts.
+  - `--tls-verify` - skip TLS verification for insecure registries.
+
+- `additionalPushOptions`
+  - `--authfile` - path to the authentication file (usually created using the `buildah login`).
+  - `--cert-dir` - path to directory with certificates that should be used for connecting to registries.
+  - `--creds` - the `[username[:password]]` to use to authenticate with the registry if required.
+  - `--retry` - number of times to retry in case of failure during image pull (the base one).
+  - `--retry-delay` - duration between retry attempts.
+  - `--tls-verify` - skip TLS verification for insecure registries.
+
+If the user-specified Buildah options will contain forbidden options (or not known), condition with warning will be added inside the `.status` section of `KafkaConnect` resource, the `InvalidResourceException` will be thrown (logged inside the operator log) and the build will fail.
```

Comment:
This makes no sense. You either ignore them with warning. Or you fail the reconciliation. Not both things. Failing the reconciliation will take care of the visibility without any additional warnings in logs or status.

What do we do for Kaniko? Fail everything or ignore with a warning?

Reply:
I checked the validation method for Kaniko and there is just the exception:

```
        if (!forbiddenOptions.isEmpty())    {
            throw new InvalidResourceException(".spec.build.additionalKanikoOptions contains forbidden options: " + forbiddenOptions);
        }
```
so it fails the reconciliation. I will change the sentence accordingly.